src/infrastructure/selectors/XmlSelectorAdapter.js

```diff
@@ -0,0 +1,44 @@
+const xpath = require('xpath');
+const { XMLSerializer, DOMParser } = require('xmldom');
+const DataSelectorPort = require('../../domain/common/port/DataSelectorPort');
+
+class XmlSelectorAdapter extends DataSelectorPort {
+  constructor() {
+    super();
+    this.domParser = new DOMParser();
+    this.xmlSerializer = new XMLSerializer();
+  }
+
+  canHandle(contentType) {
+    return ['xml', 'html'].includes(contentType);
+  }
+
+  select(data, path) {
+    const doc = this.domParser.parseFromString(data);
+    this.xmlSerializer.serializeToString(doc); // this throw an error when doc is invalid xml
+
+    if (path === null) {
+      return doc.toString();
+    }
+
+    const selected = xpath.select(path, doc);
+
+    if (selected instanceof Array && selected.length === 0) {
```

Comment:
This function has large cyclomatic complexity and 6 return points - it's hard to read, debug and modify. I suggest either using if/else and single return value or delegating responsibility of wrapping data in `<result>`/`<resultlist>` tags to another class.

Reply:
I agree but it would be made in separate PR #20.